### PR TITLE
Release3 can update

### PR DESCRIPTION
**Pull request by brgruber92**
Can Update ready for tests

### CODE REVIEW

File: src/taskpane/model.js [hunk 1]
```diff
@@ -110,6 +111,8 @@ var templateFields = {
         { field: "OwnerId", name: "Owner", type: params.fieldType.user },
         { field: "ComponentId", name: "Component", type: params.fieldType.component },
         { field: "CreationDate", name: "Creation Date", type: params.fieldType.date },
+        { field: "ConcurrencyDate", name: "Concurrency Date", type: params.fieldType.text ,isReadOnly: true, isHidden: true },
+        { field: "IndentLevel", name: "Indent Level", type: params.fieldType.text , isReadOnly: true },
```
Comment (simonhbor) [inline src/taskpane/model.js]: IndentLevel should also have isHidden: true

File: src/taskpane/model.js [hunk 2]
```diff
@@ -144,6 +147,7 @@ var templateFields = {
         { field: "ResourceCount", name: "Resources", type: params.fieldType.num },
         { field: "DaysNonWorking", name: "Non Working Days", type: params.fieldType.int },
         { field: "CreationDate", name: "Creation Date", type: params.fieldType.date },
+        { field: "ConcurrencyDate", name: "Concurrency Date", type: params.fieldType.text ,isReadOnly: true },
```
Comment (simonhbor) [inline src/taskpane/model.js]: should also have isHidden: true

File: src/taskpane/model.js [hunk 3]
```diff
@@ -192,12 +196,18 @@ var templateFields = {
...
-        { field: "EndDate", name: "End Date", type: params.fieldType.date },
+        { field: "EndDate", name: "End Date", type: params.fieldType.date},
         { field: "EstimatedEffort", name: "Estimated Effort (in mins)", type: params.fieldType.int },
         { field: "ActualEffort", name: "Actual Effort (in mins)", type: params.fieldType.int },
+        { field: "ProjectedEffort", name: "Projected Effort (in mins)", type: params.fieldType.int },
         { field: "RemainingEffort", name: "Remaining Effort (in mins)", type: params.fieldType.int },
+        { field: "RequirementId", name: "RequirementId", type: params.fieldType.int },
+        { field: "CompletionPercent", name: "Completion Percent", type: params.fieldType.int },
+        { field: "ProjectId", name: "Project ID", type: params.fieldType.int ,isReadOnly: true },
+        { field: "ConcurrencyDate", name: "Concurrency Date", type: params.fieldType.text ,isReadOnly: true },
```
Comment (simonhbor) [inline src/taskpane/model.js]: should also have isHidden: true

File: src/taskpane/model.js [hunk 4]
```diff
@@ -192,12 +196,18 @@ var templateFields = {
                 isProjectBased: true
             }
         },
-        { field: "CreationDate", name: "Creation Date", type: params.fieldType.date },
+        { field: "CreationDate", name: "Creation Date", type: params.fieldType.date, required: true },
```
Comment (simonhbor) [inline src/taskpane/model.js]: Why is this required? The API does not need this filled in on creation

File: src/taskpane/model.js [hunk 5]
```diff
@@ -243,7 +253,9 @@ var templateFields = {
             }
         },
         { field: "CreationDate", name: "Creation Date", type: params.fieldType.date },
-        { field: "PlannedDate", name: "Planned Date", type: params.fieldType.date }
+        { field: "PlannedDate", name: "Planned Date", type: params.fieldType.date },
+        { field: "ExecutionDate", name: "Execution Date", type: params.fieldType.date,isReadOnly: true },
+        { field: "ConcurrencyDate", name: "Concurrency Date", type: params.fieldType.text ,isReadOnly: true },
```
Comment (simonhbor) [inline src/taskpane/model.js]: should also have isHidden: true

File: src/taskpane/model.js [hunk 6]
```diff
@@ -292,7 +304,9 @@ var templateFields = {
         { field: "ClosedDate", name: "Closed On", type: params.fieldType.date },
         { field: "EstimatedEffort", name: "Estimated Effort (mins)", type: params.fieldType.int },
         { field: "ActualEffort", name: "Actual Effort (mins)", type: params.fieldType.int },
-        { field: "RemainingEffort", name: "Remaining Effort (mins)", type: params.fieldType.int }
+        { field: "RemainingEffort", name: "Remaining Effort (mins)", type: params.fieldType.int },
+        { field: "CompletionPercent", name: "Completion Percent", type: params.fieldType.int },
+        { field: "ConcurrencyDate", name: "Concurrency Date", type: params.fieldType.text ,isReadOnly: true },
```
Comment (simonhbor) [inline src/taskpane/model.js]: should also have isHidden: true

File: src/taskpane/model.js [hunk 7]
```diff
@@ -292,7 +304,9 @@ var templateFields = {
         { field: "ClosedDate", name: "Closed On", type: params.fieldType.date },
         { field: "EstimatedEffort", name: "Estimated Effort (mins)", type: params.fieldType.int },
         { field: "ActualEffort", name: "Actual Effort (mins)", type: params.fieldType.int },
-        { field: "RemainingEffort", name: "Remaining Effort (mins)", type: params.fieldType.int }
+        { field: "RemainingEffort", name: "Remaining Effort (mins)", type: params.fieldType.int },
+        { field: "CompletionPercent", name: "Completion Percent", type: params.fieldType.int },
```
Comment (simonhbor) [inline src/taskpane/model.js]: I think this is a calculated field: is it required for PUTs? What happens if it is left out? In other words, if we can I think we should not include this field

File: src/taskpane/model.js [hunk 8]
```diff
@@ -192,12 +196,18 @@ var templateFields = {
...
+        { field: "CreationDate", name: "Creation Date", type: params.fieldType.date, required: true },
         { field: "StartDate", name: "Start Date", type: params.fieldType.date },
-        { field: "EndDate", name: "End Date", type: params.fieldType.date },
+        { field: "EndDate", name: "End Date", type: params.fieldType.date},
         { field: "EstimatedEffort", name: "Estimated Effort (in mins)", type: params.fieldType.int },
         { field: "ActualEffort", name: "Actual Effort (in mins)", type: params.fieldType.int },
+        { field: "ProjectedEffort", name: "Projected Effort (in mins)", type: params.fieldType.int },
         { field: "RemainingEffort", name: "Remaining Effort (in mins)", type: params.fieldType.int },
+        { field: "RequirementId", name: "RequirementId", type: params.fieldType.int },
+        { field: "CompletionPercent", name: "Completion Percent", type: params.fieldType.int },
```
Comment (simonhbor) [inline src/taskpane/model.js]: This should be a calculated field. Can we exclude this field from Excel?

File: src/taskpane/model.js [hunk 9]
```diff
@@ -341,7 +355,10 @@ var templateFields = {
                 isProjectBased: true
             }
         },
-        { field: "ComponentIds", name: "Test Case Component", type: params.fieldType.component, isMulti: true }
+        { field: "ComponentIds", name: "Test Case Component", type: params.fieldType.component, isMulti: true },
+        { field: "CreationDate", name: "Test Case Creation Date", type: params.fieldType.text ,isReadOnly: true },
+        { field: "ConcurrencyDate", name: "Test Case Conc. Date", type: params.fieldType.text ,isReadOnly: true },
+        { field: "ConcurrencyDate", name: "Test Step Conc. Date", type: params.fieldType.text, isReadOnly: true , isSubTypeField: true }
```
Comment (simonhbor) [inline src/taskpane/model.js]: Should have isHidden: true

File: src/taskpane/model.js [hunk 10]
```diff
@@ -391,6 +408,8 @@ var templateFields = {
         { field: "CreationDate", name: "Creation Date", type: params.fieldType.date },
         { field: "ClosedDate", name: "Closed Date", type: params.fieldType.date },
         { field: "ReviewDate", name: "Review Date", type: params.fieldType.date },
+        { field: "RiskExposure", name: "Risk Exposure", type: params.fieldType.int, isReadOnly: true },
+        { field: "ConcurrencyDate", name: "Concurrency Date", type: params.fieldType.text ,isReadOnly: true },
```
Comment (simonhbor) [inline src/taskpane/model.js]: should have isHidden: true

File: src/taskpane/taskpane.js [hunk 11]
```diff
@@ -11,15 +11,11 @@ var model = new Data();
 var uiSelection = new tempDataStore();
 
 // if devmode enabled, set the required fields and show the dev button
-var devMode = false;
+var devMode = true;
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: this should be set to false

File: src/taskpane/taskpane.js [hunk 12]
```diff
@@ -418,18 +418,24 @@ function populateProjects(projects) {
 
 // manage the switching of the UI off the login screen on succesful login and retrieval of projects
 function showMainPanel(type) {
+  //artifactUpdateUI(false);
   setDropdown("select-product", model.projects, "Select a product");
   setDropdown("select-artifact", params.artifacts, "Select an artifact");
+  model.projects
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: think this line should be deleted as is here in error

File: src/taskpane/taskpane.js [hunk 13]
```diff
@@ -472,6 +478,8 @@ function logout(shouldLogout) {
 
 
 function changeProjectSelect(e) {
+  artifactUpdateUI(1);
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: Please add a comment explaining what this code is meant to do.
Also - please avoid "magic numbers" like this. Instead make a global object at the very top of the file which acts as enum. Then use that here instead of the one.
EG const modeEnum = { update: 1}; then here instead of 1 write modeEnum.update

File: src/taskpane/taskpane.js [hunk 14]
```diff
@@ -528,7 +536,48 @@ function getTemplateFromProjectId(user, projectId, artifact) {
 }
 
 
+function artifactUpdateUI(mode) {
+  if (mode == 1)
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: see above - do not use magic numbers but a reference to an enum defined elswhere. This makes the code more robust to future change and also easier to read and understand.
Please also add a comment about what is going on

File: src/taskpane/taskpane.js [hunk 15]
```diff
@@ -528,7 +536,48 @@ function getTemplateFromProjectId(user, projectId, artifact) {
...
 
+function artifactUpdateUI(mode) {
+  if (mode == 1) 
+  { //when selecting a new project
+    document.getElementById("main-guide-1").classList.remove("pale");
+    document.getElementById("main-guide-2").classList.add("pale");
+    document.getElementById("main-guide-3").classList.add("pale");
+    document.getElementById("btn-fromSpira").disabled = true;
+    document.getElementById("btn-updateToSpira").disabled = true;
+    //document.getElementById("select-artifact").selectedIndex = "0";
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: should delete this line

File: src/taskpane/taskpane.js [hunk 16]
```diff
@@ -528,7 +536,48 @@ function getTemplateFromProjectId(user, projectId, artifact) {
...
+  }
+
+  if (mode == 2)
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: again please use the new enum object, and not a magic number. Also add a comment about this particular case
It is not essential, but you can use an else if instead of an if. Currently the code will check through all if blocks, but really it should stop after the mode has been matched. You could also use a switch, because you are checking whether a parameter matches a specific value

File: src/taskpane/taskpane.js [hunk 17]
```diff
@@ -528,7 +536,48 @@ function getTemplateFromProjectId(user, projectId, artifact) {
...
+    
+    /*if(document.getElementById("select-artifact").selectedIndex == 0)
+    {
+      document.getElementById("main-guide-1").classList.remove("pale");
+      document.getElementById("main-guide-2").classList.add("pale");
+      document.getElementById("btn-fromSpira").disabled = true;
+    }*/
+  }
+
+  if (mode == 3)
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: see above re magic number and else if / switch

File: src/taskpane/taskpane.js [hunk 18]
```diff
@@ -528,7 +536,48 @@ function getTemplateFromProjectId(user, projectId, artifact) {
...
+    
+    document.getElementById("main-guide-3").classList.add("pale");
+    document.getElementById("btn-updateToSpira").disabled = true;
+    
+    /*if(document.getElementById("select-artifact").selectedIndex == 0)
+    {
+      document.getElementById("main-guide-1").classList.remove("pale");
+      document.getElementById("main-guide-2").classList.add("pale");
+      document.getElementById("btn-fromSpira").disabled = true;
+    }*/
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: delete these lines if they are not being used

File: src/taskpane/taskpane.js [hunk 19]
```diff
@@ -528,7 +536,48 @@ function getTemplateFromProjectId(user, projectId, artifact) {
...
+  }
+
+  if (mode == 3) 
+  { //when clicking from-Spira button
+    document.getElementById("btn-updateToSpira").disabled = false;
+  }
+
+  if (mode == 4)
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: see above re magic number and else if / switch

File: src/taskpane/taskpane.js [hunk 20]
```diff
@@ -528,7 +536,48 @@ function getTemplateFromProjectId(user, projectId, artifact) {
...
+    document.getElementById("main-guide-2").classList.remove("pale");
+    document.getElementById("btn-fromSpira").disabled = false;
+    document.getElementById("main-guide-3").classList.add("pale");
+    document.getElementById("btn-updateToSpira").disabled = true;
+  }
+}
+
+
 function changeArtifactSelect(e) {
+  artifactUpdateUI(2);
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: add a comment and do not use a magic number

File: src/taskpane/taskpane.js [hunk 21]
```diff
@@ -562,6 +609,7 @@ function manageTemplateBtnState() {
   // initially disable the button, because required API calls not completed
   document.getElementById("btn-toSpira").disabled = true;
   document.getElementById("btn-fromSpira").disabled = true;
+  //document.getElementById("btn-updateToSpira").disabled = true;
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: delete this line

File: src/taskpane/taskpane.js [hunk 22]
```diff
@@ -661,17 +725,22 @@ function getFromSpiraAttempt() {
     //if no template - then get the template
     createTemplateAttempt();
   }
+  artifactUpdateUI(3);
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: add a comment explaining why this code is here and do not use a magic number

File: src/taskpane/taskpane.js [hunk 23]
```diff
@@ -661,17 +725,22 @@ function getFromSpiraAttempt() {
...
-
-
 function getFromSpiraComplete(log) {
   if (devMode) console.log(log);
   //if array (which holds error responses) is present, and errors present
   if (log && log.errorCount) {
     errorMessages = log.entries
       .filter(function (entry) { return entry.error; })
       .map(function (entry) { return entry.message; });
+      //artifactUpdateUI(false);
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: delete this line

File: src/taskpane/taskpane.js [hunk 24]
```diff
@@ -661,17 +725,22 @@ function getFromSpiraAttempt() {
...
+  }
+  else
+  {
+    //artifactUpdateUI(true);
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: delete this line

File: src/taskpane/taskpane.js [hunk 25]
```diff
@@ -1196,7 +1288,7 @@ function templateLoaderSuccess(response) {
 
   // de-emphasise the explanatory message - we have to put it in a set timeout because there is already a delay set for other checks that affect the view state of this element
   setTimeout(function() {
-    document.getElementById("main-guide-2").classList.add("pale");
+    //document.getElementById("main-guide-2").classList.add("pale");
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: why has this line now been commented out?  what has changed for this UI change to not be used anymore?
If this code is no longer required, then the entire setTimeout is not needed anymore and should be removed.

File: src/taskpane/taskpane.js [hunk 26]
```diff
@@ -1224,11 +1316,13 @@ function errorPopUp(type, err) {
     console.log(err);
   } else {
     msOffice.error(type, err);
+    artifactUpdateUI(4);
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: again, add a comment and do not use a magic number

File: src/taskpane/taskpane.js [hunk 27]
```diff
@@ -1224,11 +1316,13 @@ function errorPopUp(type, err) {
     console.log(err);
```
Comment (simonhbor) [inline src/taskpane/taskpane.js:1308]: You need to add the same line as for Excel here: "artifactUpdateUI(4);" (corrected with comment and no magic number)

File: src/taskpane/taskpane.js [hunk 28]
```diff
@@ -1224,11 +1316,13 @@ function errorPopUp(type, err) {
...
   } else {
     msOffice.error(type, err);
+    artifactUpdateUI(4);
     console.error("SpiraPlan Import/Export Plugin encountered an error:", err.status ? err.status : "", err.response ? err.response.text : "", err.description ? err.description : "")
     console.info("SpiraPlan Import/Export Plugin: full error is... ", err)
   }
   hideLoadingSpinner();
 }
+artifactUpdateUI(4);
```
Comment (simonhbor) [inline src/taskpane/taskpane.js]: Why is this line needed? It is outside of the if block so will always run - so it will run once if Google/Excel then a second time outside of the if block. Think it should be deleted, or clearly explained why it needs to runs twice

File: src/taskpane/server.js [hunk 29]
```diff
@@ -558,23 +588,85 @@ function postArtifactToSpira(entry, user, projectId, artifactTypeId, parentId) {
...
     // TEST SETS
     case ART_ENUMS.testSets:
-    postUrl = API_PROJECT_BASE + projectId + '/test-sets?';
-    break;
+      postUrl = API_PROJECT_BASE + projectId + '/test-sets?';
+      break;
   }
 
+  //console.log('POST URL ' + postUrl);
+  //console.log('POST JSON_body ' + JSON.stringify(JSON_body));
```
Comment (simonhbor) [inline src/taskpane/server.js]: delete the commented out code

File: src/taskpane/server.js [hunk 30]
```diff
@@ -1128,14 +1220,27 @@ function contentFormattingSetter(sheet, model) {
 
     // protect column
     // read only fields - ie ones you can get from Spira but not create in Spira (as with IDs - eg task component)
-    if (model.fields[i].unsupported || model.fields[i].isReadOnly) {
+    if ((model.fields[i].unsupported || model.fields[i].isReadOnly) && !model.fields[i].isHidden) {
```
Comment (simonhbor) [inline src/taskpane/server.js]: I see how your code is working but it is a bit confusing. You should split the protectColumn function into two - each does a specific function and only one
1. protectColumn (as it was before without the extra param)
2. hideColumn that ONLY hides the column.
 
So for this specific line of code put it back to "if (model.fields[i].unsupported || model.fields[i].isReadOnly)"

File: src/taskpane/server.js [hunk 31]
```diff
@@ -1128,14 +1220,27 @@ function contentFormattingSetter(sheet, model) {
...
       }
-      
+
+      protectColumn(
+        sheet,
+        columnNumber,
+        nonHeaderRows,
+        model.colors.bgReadOnly,
+        warning,
+        false
```
Comment (simonhbor) [inline src/taskpane/server.js]: so when you refactor protectColumn this line / param will no longer be needed

File: src/taskpane/server.js [hunk 32]
```diff
@@ -1128,14 +1220,27 @@ function contentFormattingSetter(sheet, model) {
...
+      );
+    }
+
+    if (model.fields[i].isHidden) {
```
Comment (simonhbor) [inline src/taskpane/server.js:1241]: add a comment above here.
Also note that both if blocks can run on a hidden field that is protected. It will protected and then hidden in two different steps. I know that may not be ideal from an Excel perspective, but will make the code cleaner and more readable.
Then the code below this calling protectColumn will be changed to hideColumn I guess

File: src/taskpane/server.js [hunk 33]
```diff
@@ -1176,6 +1283,11 @@ function protectColumn(sheet, columnNumber, rowLength, bgColor, name, hide) {
     // set the background color
     range.set({ format: { fill: { color: bgColor } } });
 
+    //hide the column, if necessary
```
Comment (simonhbor) [inline src/taskpane/server.js]: 1. This function should be refactored as already mentioned
2. That will remove the hide param and the hide code from here.
3. NOTE: your code to hide the column is not working because I think you are using the wrong API call. it should be rangeOfColumns.hidden = true - see here: https://docs.microsoft.com/en-us/office/vba/api/excel.range.hidden

File: src/taskpane/server.js [hunk 34]
```diff
@@ -1202,13 +1314,53 @@ function protectColumn(sheet, columnNumber, rowLength, bgColor, name, hide) {
 
 }
 
+// resets ID column backgroung colors to its original - used before a GET command
+// @param: sheetName - current sheet name
 
+function resetIdColors(requiredSheetName) {
+  Excel.run(function (ctx) {
+    var rangeAddress = "A2:A5000";
```
Comment (simonhbor) [inline src/taskpane/server.js]: this range should not be hard coded for 3 reasons:
1. test step IDs are in column B
2. users may have more than 5000 rows (or far fewer)
3. it is never good to hard code things like this if you can avoid it.

File: src/taskpane/server.js [hunk 35]
```diff
@@ -1202,13 +1314,53 @@ function protectColumn(sheet, columnNumber, rowLength, bgColor, name, hide) {
...
+  });
+}
 
+// removes error messages from the 1st field of each row, in case there's any. Only the IDs are preserved
+// @param: sheetData - the sheet data object
+// returns the filtered sheetData object
 
+function clearErrorMessages(sheetData, isUpdate) {
+  console.log('clearErrorMessages sheetData');
+  console.log(sheetData);
```
Comment (simonhbor) [inline src/taskpane/server.js]: these lines should be deleted

File: src/taskpane/server.js [hunk 36]
```diff
@@ -1224,7 +1376,8 @@ function protectColumn(sheet, columnNumber, rowLength, bgColor, name, hide) {
 // function that manages exporting data from the sheet - creating an array of objects based on entered data, then sending to Spira
 // @param: model - full model object from client containing field data for specific artifact, list of project users, components, etc
 // @param: fieldTypeEnums - list of fieldType enums from client params object
```
Comment (simonhbor) [inline src/taskpane/server.js:1426]: can you please add a line for the isUpdate param

File: src/taskpane/server.js [hunk 37]
```diff
@@ -1423,23 +1575,26 @@ async function sendExportEntriesExcel(entriesForExport, sheetData, sheet, sheetR
         log.parentId = getHierarchicalParentId(entriesForExport[i].indentPosition, log.entries);
       }
 
-      await manageSendingToSpira(entriesForExport[i], model.user, model.currentProject.id, artifact, fields, fieldTypeEnums, log.parentId)
+      //set the correct parentId artifact for Test Steps and property (needed for POST URL)
+      if (artifact.id == 2 && entriesForExport[i].hasOwnProperty('TestStepId')) {
```
Comment (simonhbor) [inline src/taskpane/server.js]: I am not sure what is going on here - why are you looking for an artifact with an ID of 2?
You should not hard code a field name here if you can avoid it. Try using data from the model to do whatever this code is doing more generically / dynamically

File: src/taskpane/server.js [hunk 38]
```diff
@@ -1637,17 +1789,23 @@ function setFeedbackBgColor(cell, error, field, fieldTypeEnums, artifact, colors
 // @param: field - the field specific to the cell
 // @param: fieldTypeEnums - enum information about field types
 // @param: message - relevant error message from the entry for this row
-function setFeedbackNote(cell, error, field, fieldTypeEnums, message) {
+// @param: value - original ID of the artifact to be kept in case of an error
+function setFeedbackNote(cell, error, field, fieldTypeEnums, message, value, isUpdate) {
+  console.log('field:');
+  console.dir(field);
```
Comment (simonhbor) [inline src/taskpane/server.js]: delete these lines

File: src/taskpane/server.js [hunk 39]
```diff
@@ -1637,17 +1789,23 @@ function setFeedbackBgColor(cell, error, field, fieldTypeEnums, artifact, colors
...
 // @param: message - relevant error message from the entry for this row
-function setFeedbackNote(cell, error, field, fieldTypeEnums, message) {
+// @param: value - original ID of the artifact to be kept in case of an error
+function setFeedbackNote(cell, error, field, fieldTypeEnums, message, value, isUpdate) {
+  console.log('field:');
+  console.dir(field);
   // handle entries with errors - add error notes into ID field
   if (error && field.type == fieldTypeEnums.id) {
-    return message;
+    if (value == "-1" && isUpdate && field.field != "TestCaseId" && field.field != "TestStepId") {
```
Comment (simonhbor) [inline src/taskpane/server.js]: this magic number comes from Spira I think? Can you please add a comment explaining where -1 comes from and why it is being handled in this way?
Also avoid using hard coded field names here - I don't think POSTS handle test cases with steps in this way - eg with code like "if (artifact.hasSubType && !entry.isSubType)"

File: src/taskpane/server.js [hunk 40]
```diff
@@ -1789,6 +1985,24 @@ function getHierarchicalParentId(indent, previousEntries) {
   return -1;
 }
 
+// returns the correct parentId for the relevant group position by looping back through the list of entries
+// returns -1 if no match found
```
Comment (simonhbor) [inline src/taskpane/server.js:2041]: why does it return -1 as opposed to false? What is special about -1?

File: src/taskpane/server.js [hunk 41]
```diff
@@ -1789,6 +1985,24 @@ function getHierarchicalParentId(indent, previousEntries) {
...
+// @param: i - target artifact position to look for
+// @param: artifactId - artifact type
+function getAssociationParentId(entriesForExport, i, artifactId) {
+  // if there is no indent/ set to initial indent we return out immediately 
+  if (artifactId == 2 || artifactId == 7) {
+
+    for (i; i > 0; i--) {
+
+      if (entriesForExport[i - 1].hasOwnProperty('TestCaseId')) {
+        return entriesForExport[i - 1].TestCaseId;
```
Comment (simonhbor) [inline src/taskpane/server.js]: you should not be hard coding a specific field here - this is a generic function, but looks like it is actually only about test cases. This is very confusing. I suggest you make the function be more generic and handle all cases like test cases, not just test cases

File: src/taskpane/server.js [hunk 42]
```diff
@@ -1926,15 +2140,21 @@ function createEntryFromRow(row, model, fieldTypeEnums, artifactIsHierarchical,
 
   //we need to turn an array of values in the row into a validated object
   for (var index = 0; index < row.length; index++) {
-
+    var skipField = false;
+    if ((isUpdate && fields[index].type == fieldTypeEnums.id) || (isUpdate && fields[index].type == fieldTypeEnums.subId)) {
+      var skipField = false;
+    }
```
Comment (simonhbor) [inline src/taskpane/server.js]: This if block sets a field already false to false. What it does do however is make sure the else if blocks do not run .
So suggest you change the line "var skipField = false;" to simply "//do nothing"

File: src/taskpane/server.js [hunk 43]
```diff
@@ -1926,15 +2140,21 @@ function createEntryFromRow(row, model, fieldTypeEnums, artifactIsHierarchical,
...
-    if (fieldsToFilter == FIELD_MANAGEMENT_ENUMS.standard && fields[index].isSubTypeField) {
+    else if (fieldsToFilter == FIELD_MANAGEMENT_ENUMS.standard && fields[index].isSubTypeField) {
       // skip the field
+      skipField = true;
+
     } else if (fieldsToFilter == FIELD_MANAGEMENT_ENUMS.subType && !(fields[index].isSubTypeField || fields[index].isTypeAndSubTypeField)) {
       // skip the field
-
+      skipField = true;
       // in all other cases add the field
```
Comment (simonhbor) [inline src/taskpane/server.js]: should this comment be moved down a line? I am not sure what it is referring to

**Review by simonhbor — COMMENTED** [review]
The code is looking good and the logic makes sense, as does adding the isUpdate flag as a way of managing the logic well.
I have added a number of comments, some for small changes others for function refactors. Let me know if you have any questions

File: src/taskpane/model.js (hunk 4 above)
Comment (brgruber92) [reply src/taskpane/model.js]: That's true - that shouldn't be required.

File: src/taskpane/model.js (hunk 7 above)
Comment (brgruber92) [reply src/taskpane/model.js]: This is not a required field for PUT, so I removed it from the model.

File: src/taskpane/model.js (hunk 8 above)
Comment (brgruber92) [reply src/taskpane/model.js]: Yes, removing that don't cause any problems

File: src/taskpane/taskpane.js (hunk 16 above)
Comment (brgruber92) [reply src/taskpane/taskpane.js]: Changed to switch + comments

File: src/taskpane/taskpane.js (hunk 25 above)
Comment (brgruber92) [reply src/taskpane/taskpane.js]: This part seems to inactivate the buttons of the "main-guide-2" part of the task pane - that's not necessary anymore, since it's been handled in other parts of the code. I removed the function.

File: src/taskpane/taskpane.js (hunk 28 above)
Comment (brgruber92) [reply src/taskpane/taskpane.js]: It doesn't need to run twice - that was a mistake. Already fixed.

File: src/taskpane/server.js (hunk 32 above)
Comment (brgruber92) [reply src/taskpane/server.js:1241]: That's true. I changed as specified.

File: src/taskpane/server.js (hunk 33 above)
Comment (brgruber92) [reply src/taskpane/server.js]: I fixed the Hide function using range.columnHidden = true;

File: src/taskpane/server.js (hunk 39 above)
Comment (brgruber92) [reply src/taskpane/server.js]: Comments added and enumeration implemented to make the code easier to understand and more generic

File: src/taskpane/server.js (hunk 40 above)
Comment (brgruber92) [reply src/taskpane/server.js:2041]: There's nothing special actually - I just wanted to keep the original 'log' value, as it was originally created as -1. Also, the idea is to provide an invalid ID (integer), so I choose -1. I think false won't work.

File: src/taskpane/server.js (hunk 34 above)
Comment (brgruber92) [reply src/taskpane/server.js:1322]: I'll need help to find a better place for this function because where it is right now, the only argument I have to pass to it is the sheet name - in order to determine what should be the IDs range we are modifying, I'll need extra information.

File: src/taskpane/server.js (hunk 41 above)
Comment (brgruber92) [reply src/taskpane/server.js]: Changes done - the code now uses generic objects and enumerations

File: src/taskpane/server.js (hunk 34 above)
Comment (brgruber92) [reply src/taskpane/server.js]: I completely redraw this function - now it gets the whole range of old data and restores the background colors of each column to their original - not only the IDs. They're not hard-coded values anymore.

File: src/taskpane/model.js (hunk 7 above)
Comment (simonhbor) [reply src/taskpane/model.js]: It looks like the line is still there - suggest just deleting it

**Review by simonhbor — CHANGES_REQUESTED** [review]
Think there is just one line to delete in model.js. Rest looks good
